src/python/pants/backend/python/util_rules/local_dists.py

```diff
@@ -150,10 +150,8 @@ async def build_local_dists(
     for source in request.sources.source_files.files:
         if source not in unrooted_files_set:
             for source_root in source_roots:
-                if (
-                    source.startswith(source_root)
-                    and os.path.relpath(source, source_root) in provided_files
-                ):
+                source_relpath = fast_relpath_optional(source, source_root)
```

Comment:
It might be good of source_roots were non-str / non-PathLike objects so attempts to use `os` APIs failed, forcing you to convert them to a PathLike 1st. That might be enough pause to see new, perf appropriate API's hung off the source root type.

Reply:
Mm... yea. I think that for a bit there we thought that `Path`/`PurePath` were going to be this better API, but it unfortunately seems that that isn't the case.

The challenge is always how widespread these APIs are: there is a huge surface area in the stdlib that takes either `str`/`Path`, so any custom path type has a high bar to be worth the extra boilerplate of conversions at the boundary. And we don't have `Deref` like on the Rust side.

But yea, maybe if it were isolated to a SourceRoot type... although that limits the upside of the custom type as well.